cartridge/upload.lua

```diff
@@ -0,0 +1,322 @@
+--- Spread the data across instances in a network-efficient manner.
+--
+-- (**Added** in v2.4.0-43)
+--
+-- @module cartridge.upload
+-- @local
+
+local log = require('log')
+local fio = require('fio')
+local errno = require('errno')
+local fiber = require('fiber')
+local checks = require('checks')
+local errors = require('errors')
+local digest = require('digest')
+local msgpack = require('msgpack')
+local pool = require('cartridge.pool')
+local utils = require('cartridge.utils')
+
+local vars = require('cartridge.vars').new('cartridge.upload')
+
+local UploadError = errors.new_class('UploadError')
+
+--- The uploaded data.
+-- `{[upload_id] = data}`
+-- @table inbox
+vars:new('inbox', {})
+
+-- All four functions begin / transmit / finish / cleanup do yield.
+-- Since they are called over netbox (each in an individual fiber),
+-- there may be a situation (usually abnormal), when cleanup starts
+-- before previous stages finish.
+vars:new('upload_fibers', {})
+
+-- Temporary directory used for saving files during upload.
+vars:new('upload_prefix', '/tmp')
+
+local function set_upload_prefix(prefix)
+    vars.upload_prefix = prefix
+end
+
+local function get_upload_prefix()
+    return vars.upload_prefix
+end
+
+local function get_upload_path(upload_id)
+    checks('string')
+    return fio.pathjoin(
+        vars.upload_prefix,
+        'cartridge-upload.' .. upload_id
+    )
+end
+
+-- The upload starts by creating a shared resource - a directory named
+-- after the `upload_id`. The instances who can create it before
+-- others are the transmitters. The others wait patiently till the
+-- `upload_finish()`.
+local function upload_begin(upload_id)
+    checks('string')
+
+    vars.upload_fibers[upload_id] = fiber.self()
+
+    local upload_path = get_upload_path(upload_id)
+    local ok, err = utils.mktree(fio.dirname(upload_path))
+    fiber.testcancel()
```

Comment:
`testcancel` raises. You won't perform `vars.upload_fibers[upload_id] = nil` if this function raises an error.

Reply:
It's ok. In this case the `upload_fibers` will be cleaned by the `cleanup` stage which cancelled the fiber.

Reply:
Ok, but it's not so obvious, please add a comment.

Reply:
However, in fact we can't guarantee that cleanup stage will be executed someday in case of temporary problems with network.

Reply:
> Ok, but it's not so obvious, please add a comment.

Done

> However, in fact we can't guarantee that cleanup stage will be executed someday in case of temporary problems with network.

The necessity of persisting `upload_fibers` is to avoid a race between `begin` and `cleanup`. The necessity of cleaning up the `upload_fibers` is to avoid canceling an innocent fiber that belongs to the iproto fiber pool and has already been reused by that time.

Reply:
To sum up, if the cleanup stage wasn't called for some reason, the garbage will remain in `/tmp` and in lua context. And it's ok for now.